ddtrace/internal/import_hooks.py

```diff
@@ -0,0 +1,361 @@
+    if module_name and module_name not in sys.modules:
+        return exec_and_call_hooks(module_name, wrapped, args, kwargs)
+
+    return wrapped(*args, **kwargs)
+
+
+# Keep track of whether we have patched or not
+_patched = False
+
+
+def _patch():
+    # Only patch once
+    global _patched
+    if _patched:
+        return
+
+    # 3.x
+    if PY3:
+        # 3.4: https://github.com/python/cpython/blob/3.4/Lib/importlib/_bootstrap.py#L2207-L2231
+        # 3.5: https://github.com/python/cpython/blob/3.5/Lib/importlib/_bootstrap.py#L938-L962
+        # 3.6: https://github.com/python/cpython/blob/3.6/Lib/importlib/_bootstrap.py#L936-L960
+        # 3.7: https://github.com/python/cpython/blob/3.7/Lib/importlib/_bootstrap.py#L948-L972
+        # 3.8: https://github.com/python/cpython/blob/3.8/Lib/importlib/_bootstrap.py#L956-L980
+        # DEV: Python 3 has multiple entrypoints for importing a module, but `_find_and_load_unlocked`
+        #   is a common base/required function needed by anyone to import a module.
+        #   e.g. `__import__` which calls `importlib._bootstrap._find_and_load()`
+        #        `importlib.__import__/importlib._bootstrap.__import__` which calls `importlib._bootstrap._gcd_import()`
+        #        `importlib.import_module` which calls `importliob._bootstrap._gcd_import()`
+        wrapt.wrap_function_wrapper('importlib._bootstrap', '_find_and_load_unlocked', wrapped_find_and_load_unlocked)
+
+        # 3.4: https://github.com/python/cpython/blob/3.4/Lib/importlib/__init__.py#L115-L156
+        # 3.5: https://github.com/python/cpython/blob/3.5/Lib/importlib/__init__.py#L132-L173
+        # 3.6: https://github.com/python/cpython/blob/3.6/Lib/importlib/__init__.py#L132-L173
+        # 3.7: https://github.com/python/cpython/blob/3.7/Lib/importlib/__init__.py#L133-L176
+        # 3.8: https://github.com/python/cpython/blob/3.8/Lib/importlib/__init__.py#L133-L176
+        wrapt.wrap_function_wrapper('importlib', 'reload', wrapped_reload)
+
+    # 2.7
+    # DEV: Slightly more direct approach of patching `__import__` and `reload` functions
+    elif sys.version_info >= (2, 7):
+        # https://github.com/python/cpython/blob/2.7/Python/bltinmodule.c#L35-L68
+        __builtins__['__import__'] = wrapt.FunctionWrapper(__builtins__['__import__'], wrapped_import)
+
+        # https://github.com/python/cpython/blob/2.7/Python/bltinmodule.c#L2147-L2160
+        __builtins__['reload'] = wrapt.FunctionWrapper(__builtins__['reload'], wrapped_reload)
+
+    # Update after we have successfully patched
+    _patched = True
+
+
+# DEV: This is called at the end of this module to ensure we always patch
+def patch():
+    """
+    Patch Python import system, enabling import hooks
+    """
+    # This should never cause their application to not load
+    try:
+        _patch()
+    except Exception:
+        log.debug('Failed to patch module importing, import hooks will not work', exc_info=True)
```

Comment:
I don't think that's a good UX here. If `patch` fails and you don't have debug log enabled, nothing will work as expected and you'll never know… this is awful.
Leaving the exception be propagated is probably safer. It says "hey that does not work" so the user has to fix it or stop calling it anyway. (or can still wrap with `try/except` if they know what they do).

Reply:
I don't think letting this exception raise up is a great idea, but we can definitely increase the visibility on this log line.

I can change to `log.warning` or `log.error`, as people should know something is wrong, but we should never stop their application from running.

on `unpatch`, realistically this will probably never get called. However, if the user manually calls this and it fails then their application likely won't do what they intended, so we should raise.

Reply:
The question is, what's the best thing to do:

1. Let the user know patching failed and therefore that the tracer does not work. Realistically this happens at startup so the user knows when testing or before deploying its app.

2. Fail more or less silently (depending on the user seeing the log) and run an application that is possible half-patched/half-broken/half-tracing with consequences we might not understand.

(I'm not saying I have the right answer here, I'm just pondering options)

Reply:
Yeah, I guess my reasoning here is that if something works in dev, but doesn't in prod, do we want to be the reason someone's application doesn't actually start?

I think my instinct (with regards to `dd-trace-py` specifically) has been to inform them as loudly as we can, but don't fail. My concern here is that if we cause someone's production deployment to fail they will probably just turn us off and then we have to rebuild the trust to get them to re-enable it. 

If we fail in a way that just disables our instrumentation, that is an inconvenience for sure, but we just need to resolve as "we fixed the problem", not "we won't cause your app to crash again".


Now again... is this the _best_ approach? /shrug but how I have been thinking about it.

Reply:
Right. I might also be too pessimistic about option 2 now that I rethink about it.

I think it's alright we can start with that and revisit if we need. We just need to be sure that users will go through their log if they don't see the tracer working.